lib/charms/vault_k8s/v0/vault_kv.py

```diff
@@ -0,0 +1,528 @@
+
+
+class VaultKvProvides(ops.Object):
+    """Class to be instanciated by the providing side of the relation."""
+
+    on = VaultKvProviderEvents()
+
+    def __init__(
+        self,
+        charm: ops.CharmBase,
+        relation_name: str,
+    ) -> None:
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.relation_name = relation_name
+        self.framework.observe(
+            self.charm.on[relation_name].relation_changed,
+            self._on_relation_changed,
+        )
+
+    def _on_relation_changed(self, event: ops.RelationChangedEvent):
+        """Handle client changed relation.
+
+        This handler will emit a new_vault_kv_client_attached event if at least one unit data is
+        valid.
+        """
+        if event.app is None:
+            logger.debug("No remote application yet")
+            return
+
+        app_data = dict(event.relation.data[event.app])
+
+        any_valid = False
+        for unit in event.relation.units:
+            if not is_requirer_data_valid(app_data, dict(event.relation.data[unit])):
+                logger.debug("Invalid data from unit %r", unit.name)
+                continue
+            any_valid = True
+
+        if any_valid:
+            self.on.new_vault_kv_client_attached.emit(
+                event.relation.id,
+                event.relation.name,
+                event.relation.data[event.app]["mount_suffix"],
+            )
+
+    def set_vault_url(self, relation: ops.Relation, vault_url: str):
+        """Set the vault_url on the relation."""
+        if not self.charm.unit.is_leader():
+            return
+
+        relation.data[self.charm.app]["vault_url"] = vault_url
+
+    def set_ca_certificate(self, relation: ops.Relation, ca_certificate: str):
+        """Set the ca_certificate on the relation."""
+        if not self.charm.unit.is_leader():
+            return
+
+        relation.data[self.charm.app]["ca_certificate"] = ca_certificate
+
+    def set_mount(self, relation: ops.Relation, mount: str):
+        """Set the mount on the relation."""
+        if not self.charm.unit.is_leader():
+            return
+
+        relation.data[self.charm.app]["mount"] = mount
+
+    def set_unit_credentials(self, relation: ops.Relation, nonce: str, secret: ops.Secret):
+        """Set the unit credentials on the relation."""
+        if not self.charm.unit.is_leader():
+            return
+
+        credentials = self.get_credentials(relation)
+        if secret.id is None:
+            logger.debug(
+                "Secret id is None, not updating the relation '%s:%d' for nonce %r",
+                relation.name,
+                relation.id,
+                nonce,
+            )
+            return
+        credentials[nonce] = secret.id
+        relation.data[self.charm.app]["credentials"] = json.dumps(credentials, sort_keys=True)
+
+    def remove_unit_credentials(self, relation: ops.Relation, nonce: Union[str, List[str]]):
```

Comment:
This type hint would be more generic and would let us send `Sets` directly.
```suggestion
    def remove_unit_credentials(self, relation: ops.Relation, nonce: Union[str, Iterable[str]]):
```